src/Simplify.cpp

```diff
@@ -137,6 +147,83 @@ class Simplify : public IRMutator {
+    bool const_int_bounds(Expr e, int64_t *min_val, int64_t *max_val) {
+        if (!no_overflow_scalar_int(e.type())) {
+            return false;
+        }
+
+        int64_t i;
+        if (const_int(e, &i)) {
+            *min_val = *max_val = i;
+            return true;
+        } else if (const Variable *v = e.as<Variable>()) {
+            if (bounds_info.contains(v->name)) {
+                pair<int64_t, int64_t> b = bounds_info.get(v->name);
+                *min_val = b.first;
+                *max_val = b.second;
+                return true;
+            }
+        } else if (const Add *add = e.as<Add>()) {
+            int64_t min_a, min_b, max_a, max_b;
+            if (const_int_bounds(add->a, &min_a, &max_a) &&
+                const_int_bounds(add->b, &min_b, &max_b)) {
+                *min_val = min_a + min_b;
+                *max_val = max_a + max_b;
+                return true;
+            }
+        } else if (const Sub *sub = e.as<Sub>()) {
+            int64_t min_a, min_b, max_a, max_b;
+            if (const_int_bounds(sub->a, &min_a, &max_a) &&
+                const_int_bounds(sub->b, &min_b, &max_b)) {
+                *min_val = min_a - max_b;
+                *max_val = max_a - min_b;
+                return true;
+            }
+        } else if (const Max *max = e.as<Max>()) {
+            int64_t min_a, min_b, max_a, max_b;
+            if (const_int_bounds(max->a, &min_a, &max_a) &&
+                const_int_bounds(max->b, &min_b, &max_b)) {
+                *min_val = std::max(min_a, min_b);
+                *max_val = std::max(max_a, max_b);
+                return true;
+            }
+        } else if (const Min *min = e.as<Min>()) {
+            int64_t min_a, min_b, max_a, max_b;
+            if (const_int_bounds(min->a, &min_a, &max_a) &&
+                const_int_bounds(min->b, &min_b, &max_b)) {
+                *min_val = std::min(min_a, min_b);
+                *max_val = std::min(max_a, max_b);
+                return true;
+            }
+        } else if (const Mul *mul = e.as<Mul>()) {
+            int64_t min_a, min_b, max_a, max_b;
+            if (const_int_bounds(mul->a, &min_a, &max_a) &&
+                const_int_bounds(mul->b, &min_b, &max_b)) {
+                int64_t t0 = min_a*min_b, t1 = min_a*max_b,
+                        t2 = max_a*min_b, t3 = max_a*max_b;
+                *min_val = std::min(std::min(t0, t1), std::min(t2, t3));
+                *max_val = std::max(std::max(t0, t1), std::max(t2, t3));
+                return true;
+            }
+        } else if (const Select *sel = e.as<Select>()) {
+            int64_t min_a, min_b, max_a, max_b;
+            if (const_int_bounds(sel->true_value, &min_a, &max_a) &&
+                const_int_bounds(sel->false_value, &min_b, &max_b)) {
+                *min_val = std::min(min_a, min_b);
+                *max_val = std::max(max_a, max_b);
+                return true;
+            }
+        }
```

Comment:
How about handling Mod? I think there's a lot of code that simplifies well if the bounds of a mod expression with a constant modulus is understood.